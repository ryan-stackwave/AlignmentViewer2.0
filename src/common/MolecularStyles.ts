/**
 * This file centralizes the location of javascript style definitions
 * for nucleotide and amino acid styles.
 */
import styles from "./MolecularStyles.module.scss";
//console.log("STYLES AS INPUT ::::: ", styles);

/**
 * Defines a single color scheme interface which consists of
 * a name, className, background alpha, and set of colors for
 * all residues (hex string format).
 *
 * @export
 * @interface IColorScheme
 */
export interface IColorScheme {
  commonName: string;
  description: string;
  backgroundAlpha: number;
  className: string;
  colors: {
    [residueCode: string]: string;
  };
}

export const ALL_AMINOACID_COLORSCHEMES = assembleColorSchemes("aa");
export const ALL_NUCLEOTIDE_COLORSCHEMES = assembleColorSchemes("nt");

/**
 * Read in class names and colors for amino acid or nucleotide
 * color schemes from the scss file.
 *
 * @param {("aa" | "nt")} moleculeType
 * @returns
 */
function assembleColorSchemes(moleculeType: "aa" | "nt") {
  return Object.keys(styles).reduce((acc, styleName) => {
    if (styleName.indexOf(moleculeType + "StyClass_") !== -1) {
      const colorSchemeName = styleName.split(moleculeType + "StyClass_")[1];
      const colorOrder = styles[
        moleculeType + "StyColorOrder_" + colorSchemeName
      ]
        .replace(/ /g, "")
        .split(","); // "build" removes spaces, but local "run" does not
      const hexValues = styles[moleculeType + "StyColors_" + colorSchemeName]
        .replace(/ /g, "")
        .split(","); // "build" removes spaces, but local "run" does not
      const description = styles[moleculeType + "StyDesc_" + colorSchemeName];
      acc.push({
        commonName: colorSchemeName,
        description: description,
        className: styles[styleName],
        backgroundAlpha: parseFloat(
          styles[moleculeType + "StyBGAlpha_" + colorSchemeName]
        ),
        colors: Object.fromEntries(
          colorOrder.map((_, i) => [colorOrder[i], hexValues[i]])
        ),
      });
    }
    return acc;
  }, [] as IColorScheme[]);
}

/**
 * Top level alignment type descriptions. There are two supported
 * alignment types: AMINOACID and NUCLEOTIDE.
 *
 * @export
 * @class AlignmentTypes
 */
export class AlignmentTypes {
  static readonly AMINOACID = new AlignmentTypes(
    "aminoacid-alignment",
    "Amino Acid Sequences",
    styles.aaAlignTypeClass,
    ALL_AMINOACID_COLORSCHEMES
  );
  static readonly NUCLEOTIDE = new AlignmentTypes(
    "nucleotide-alignment",
    "Nucleotide Sequences",
    styles.ntAlignTypeClass,
    ALL_NUCLEOTIDE_COLORSCHEMES
  );

  static list = [AlignmentTypes.AMINOACID, AlignmentTypes.NUCLEOTIDE];

  static fromKey(key: string) {
    return AlignmentTypes.list.find((at) => {
      return at.key === key;
    });
  }

  private constructor(
    public readonly key: string,
    public readonly description: string,
    public readonly className: string,
    public readonly allColorSchemes: IColorScheme[]
  ) {}
}

/**
 * This class represents different position styling modes.
 *
 * @export
 * @class PositionsToStyle
 */
export class PositionsToStyle {
  static readonly ALL = new PositionsToStyle(
    "all",
    "All",
    styles.styPosAllClass
  );
  static readonly QUERY = new PositionsToStyle(
    "query",
    "Same as Query",
    styles.styPosQueryClass
  );
  static readonly QUERY_DIFF = new PositionsToStyle(
    "query-diff",
    "Different from Query",
    styles.styPosQueryDiffClass
  );
  static readonly CONSENSUS = new PositionsToStyle(
    "consensus",
    "Same as Consensus",
    styles.styPosConsensusClass
  );
  static readonly CONSENSUS_DIFF = new PositionsToStyle(
    "consensus-diff",
    "Different from Consensus",
    styles.styPosConsensusDiffClass
  );

  static list = [
    PositionsToStyle.ALL,
    PositionsToStyle.QUERY,
    PositionsToStyle.QUERY_DIFF,
    PositionsToStyle.CONSENSUS,
    PositionsToStyle.CONSENSUS_DIFF,
  ];

  static fromKey(key: string) {
    return PositionsToStyle.list.find((at) => {
      return at.key === key;
    });
  }

  // private to disallow creating other instances of this type
  private constructor(
    public readonly key: string,
    public readonly description: string,
    public readonly className: string
  ) {}
}

export class ResidueDetailTypes {
  static readonly BOTH = new ResidueDetailTypes(
    "both",
    "Style Background and Letters",
    ""
  );
  static readonly LETTER_ONLY = new ResidueDetailTypes(
    "letter-only",
    "Only Style Letters",
    styles.transparentBackgroundClass
  );
  static readonly BACKGROUND_ONLY = new ResidueDetailTypes(
    "background-only",
    "Only Style Background",
    styles.transparentLetterClass
  );

  static list = [
    ResidueDetailTypes.BOTH,
    ResidueDetailTypes.LETTER_ONLY,
    ResidueDetailTypes.BACKGROUND_ONLY,
  ];

  static fromKey(key: string) {
    return ResidueDetailTypes.list.find((at) => {
      return at.key === key;
    });
  }

  private constructor(
    public readonly key: string,
    public readonly description: string,
    public readonly className: string
  ) {}
}

/**
 * Object to describe the style of a set of sequences (MSA, logo, etc).
 * Contains 2 key parameters:
 *     "alignmentType": should be set by implementing classes
 *     "colorScheme": describes the color scheme to apply to the alignment.
 *                    defaults to the first available from the scss file
 *     "positionsToStyle": description of which positions should be styled.
 *                         defaults to all positions.
 *
 * @interface AlignmentStyle
 */
export abstract class AlignmentStyle {
  abstract readonly alignmentType: AlignmentTypes;
  abstract colorScheme: IColorScheme;
  abstract positionsToStyle: PositionsToStyle;
  abstract residueDetail: ResidueDetailTypes;

  static fromAlignmentType(alignmentType: AlignmentTypes) {
    if (alignmentType === AlignmentTypes.AMINOACID)
      return new AminoAcidAlignmentStyle();
    return new NucleotideAlignmentStyle();
  }
}

/**
 * Object to describe an amino acid alignment style
 * @class AminoAcidAlignmentStyle
 * @implements {AlignmentStyle}
 */
export class AminoAcidAlignmentStyle implements AlignmentStyle {
  readonly allColorSchemes = ALL_AMINOACID_COLORSCHEMES;
  readonly alignmentType = AlignmentTypes.AMINOACID;

  constructor(
    public colorScheme: IColorScheme = ALL_AMINOACID_COLORSCHEMES[0],
    public positionsToStyle: PositionsToStyle = PositionsToStyle.ALL,
    public residueDetail: ResidueDetailTypes = ResidueDetailTypes.BOTH
  ) {}
}

/**
 * Object to describe an nucleotide alignment style
 *
 * @export
 * @class NucleotideAlignmentStyle
 * @implements {AlignmentStyle}
 */
export class NucleotideAlignmentStyle implements AlignmentStyle {
  readonly allColorSchemes = ALL_NUCLEOTIDE_COLORSCHEMES;
  readonly alignmentType = AlignmentTypes.NUCLEOTIDE;

  constructor(
    public colorScheme: IColorScheme = ALL_NUCLEOTIDE_COLORSCHEMES[0],
    public positionsToStyle: PositionsToStyle = PositionsToStyle.ALL,
    public residueDetail: ResidueDetailTypes = ResidueDetailTypes.BOTH
  ) {}
}

/**
 * Export globals
 */
const transparentLetterClass = styles.transparentLetterClass;
const transparentBackgroundClass = styles.transparentBackgroundClass;
export { transparentBackgroundClass, transparentLetterClass };

/**
 * Export ace helper parameters and functions.
 */
/**
 * Export a class name array that is used by the ace editor
 * mode to fill class names for each letter. The structure
 * of this string is just the list of class names, separated
 * by a period. Ace subsequently prepends ace_ to each value
 * when adding the classes to the editor.
 * @param letter
 * @param isConsensus
 * @param isQuery
 */
export function getLetterClassNamesForAce(
  letter: string,
  isConsensus: boolean,
  isQuery: boolean
) {
  const all = "." + styles.preAceLetterObjClass;
  const consensus = isConsensus ? "." + styles.preAceConsensusClass : "";
  const query = isQuery ? "." + styles.preAceQueryClass : "";

  return letter + all + consensus + query;
}

const acePrefix = styles.acePrefix;
/**
 * Get letter class names that have the final ace_ prepended
 * classnames.
 * @param letter
 * @param isConsensus
 * @param isQuery
 */
export function getFinalLetterClassNames(
  letter: string,
  isConsensus: boolean,
  isQuery: boolean
) {
  const all = " " + styles.aceLetterObjClass;
  const consensus = isConsensus ? " " + styles.aceConsensusClass : "";
  const query = isQuery ? " " + styles.aceQueryClass : "";

  return styles.acePrefix + letter + all + consensus + query;
}
