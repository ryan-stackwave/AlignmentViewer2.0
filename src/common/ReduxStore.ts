import {
  configureStore,
  ThunkAction,
  Action,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";

/*
 *
 *
 * WebGL detailed alignment view viewport location
 *
 *
 */

interface WebGLViewportState {
  pixelsFromWorldTop: number;

  //viewport: {
  //  firstSequenceIdx: number;
  //  firstSequenceWorldTop: number;
  //  residueWidth: number;
  //  residueHeight: number;
  //};
}

const initialState: WebGLViewportState = {
  pixelsFromWorldTop: 0,
};

export const webGLViewportSlice = createSlice({
  name: "webgl-viewport",
  initialState,
  reducers: {
    // Use the PayloadAction type to declare the contents of `action.payload`
    setPixelsFromWorldTop: (state, action: PayloadAction<number>) => {
      state.pixelsFromWorldTop = action.payload;
    },
  },
});

export const { setPixelsFromWorldTop } = webGLViewportSlice.actions;

/*
 *
 *
 * main store
 *
 *
 */

export const store = configureStore({
  reducer: {
    webglViewport: webGLViewportSlice.reducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
