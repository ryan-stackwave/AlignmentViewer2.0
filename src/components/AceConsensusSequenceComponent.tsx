import AceMultipleSequenceAlignmentComponent from "./AceMultipleSequenceAlignmentComponent";

export default class AceConsensusSequenceComponent extends AceMultipleSequenceAlignmentComponent {
  getText() {
    return this.props.alignment
      .getConsensus()
      .map((letter) => {
        return letter.letter;
      })
      .join("");
  }
}
