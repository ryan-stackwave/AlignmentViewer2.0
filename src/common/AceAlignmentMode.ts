/**
 * AceAlignmentMode.js
 * This file defines an alignment highlight mode form ace editor documents.
 */

import { default as ace } from "ace-builds";
import Alignment from "./Alignment";
import { getClassNamesForAce } from "./MolecularStyles";

interface IAceType {
  [key: string]: any;
}

export function defineNewAlignmentMode(mode: string, alignment: Alignment) {
  const modeSplit = mode.split("/");
  if (
    modeSplit.length !== 3 ||
    modeSplit[0] !== "ace" ||
    modeSplit[1] !== "mode"
  ) {
    throw Error("ace mode name must be of the form '/ace/mode/MODENAME'");
  }
  const modeName = mode.split("/")[2];

  (ace as IAceType).define(
    "ace/mode/" + modeName + "_highlight_rules",
    function (require: Function, exports: any, module: any) {
      var oop = require("../lib/oop");
      var TextHighlightRules = require("./text_highlight_rules")
        .TextHighlightRules;

      /**
       * Rules parse individual characters in the editor and assign classes
       * to each character. Note that characters may be grouped together
       * if they match the same rule and are side by side.
       */
      var AlignmentHighlightRules = function (this: any) {
        //We define an alignment-specific set of rules that will
        //add other tags such as whether a character matches the
        //consensus sequence.
        let rules: {
          [key: string]: {
            token: (letter: string) => void;
            regex: string;
            next: string;
          }[];
        } = {};

        for (let i = 0; i < alignment.getMaxSequenceLength(); i++) {
          const stateName = i === 0 ? "start" : "pos" + i;
          const nextStateName =
            i === alignment.getMaxSequenceLength() - 1
              ? "start"
              : "pos" + (i + 1);

          rules[stateName] = [
            {
              token: function (letter: string) {
                return getClassNamesForAce(
                  letter,
                  letter === alignment.getConsensus()[i].letter,
                  letter === alignment.getTargetSequence().sequence[i]
                );
              },
              regex: ".",
              next: nextStateName,
            },
          ];
        }
        this.$rules = rules;
      };

      oop.inherits(AlignmentHighlightRules, TextHighlightRules);

      exports.AlignmentHighlightRules = AlignmentHighlightRules;
    }
  );

  (ace as IAceType).define("ace/mode/" + modeName, function (
    require: Function,
    exports: any,
    module: any
  ) {
    var oop = require("../lib/oop");
    var TextMode = require("../mode/text").Mode;
    var Tokenizer = require("../tokenizer").Tokenizer;
    var AlignmentHighlightRules = require("ace/mode/" +
      modeName +
      "_highlight_rules").AlignmentHighlightRules;

    const Mode = function (this: any) {
      this.$tokenizer = new Tokenizer(new AlignmentHighlightRules().getRules());
    };

    oop.inherits(Mode, TextMode);

    (function () {
      // Extra logic goes here. (see below)
    }.call(Mode.prototype));

    exports.Mode = Mode;
  });
}
