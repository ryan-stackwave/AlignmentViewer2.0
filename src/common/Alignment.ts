import { generateUUIDv4 } from "./Utils";
import { defineNewAlignmentMode } from "./AceAlignmentMode";
import { Nucleotide } from "./Residues";
import { SequenceSorter } from "./AlignmentSorter";
import {
  AlignmentTypes,
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
} from "./MolecularStyles";

export interface ISequence {
  id: string;
  sequence: string;
}

/**
 * Alignment
 * This class represents a multiple sequence alignment. During initialization,
 * positional statistics are generated about the alignment and a consensus
 * sequence is determined.
 *
 * The class should generally be instantiated from subclasses (e.g., FastaAlignemtn)
 * using static methods, e.g.,
 *     FastaAlignment.fromFileContents
 *     StockholmAlignment.fromFileContents
 */
export class Alignment {
  private uuid: string;
  private aceEditorMode: string | undefined;
  private name: string;
  private predictedNT: boolean;
  private sequences: Map<SequenceSorter, ISequence[]>;
  private querySequence: ISequence;
  private positionalLetterCounts = new Map<
    number,
    { [letter: string]: number }
  >();
  private globalAlphaLetterCounts: { [letter: string]: number } = {};
  private allUpperAlphaLettersInAlignmentSorted: string[];
  private consensus: {
    sequence: ISequence;
    statistics: {
      letter: string;
      position: number;
      occurrences: number;
    }[];
  };

  /**
   * Normalize all values in an object that contains all counts such that
   * the sum of those counts equals 1.
   * @param letterCounts an object that contains the same keys, but whose
   *                     values now sum to 1.
   * @param validLetters if provided, returned object will only contain letters
   *                     that are in this array.
   */
  private static normalizeLetterCounts(
    letterCounts: { [letter: string]: number },
    validLetters?: string[]
  ) {
    const totalCounts = Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc += letterCounts[letter];
      }
      return acc;
    }, 0);

    return Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc[letter] = letterCounts[letter] / totalCounts;
      }
      return acc;
    }, {} as { [letter: string]: number });
  }

  /**
   * Create a new Alignment object. The default query sequence will
   * be set to the first sequence in sequences. This can be changed
   * by calling setQuerySequence() after creation
   * @param sequences
   */
  public constructor(name: string, sequencesAsInput: ISequence[]) {
    this.uuid = generateUUIDv4();
    this.name = name;
    this.sequences = new Map<SequenceSorter, ISequence[]>();
    this.sequences.set(SequenceSorter.INPUT, sequencesAsInput);
    this.querySequence = sequencesAsInput[0];
    this.predictedNT = true;

    //
    //generate statistics
    //
    const start = new Date();
    const allLettersInAlignment: { [key: string]: number } = {}; //all letters in the alignment

    // aggregate stats for each position and globally including
    // the number of time each amino acid occurs
    // ** this loop takes the bulk of the initialization time
    for (
      let sequenceIdx = 0;
      sequenceIdx < sequencesAsInput.length;
      sequenceIdx++
    ) {
      const seq = sequencesAsInput[sequenceIdx].sequence;
      for (let positionIdx = 0; positionIdx < seq.length; positionIdx++) {
        const letter = seq[positionIdx];
        const letterIsAlpha = letter.match(/[a-z]/i) ? true : false;
        if (
          letterIsAlpha &&
          this.predictedNT &&
          Nucleotide.fromSingleLetterCode(letter) === Nucleotide.UNKNOWN
        ) {
          this.predictedNT = false;
        }

        const position = positionIdx; // zero based
        allLettersInAlignment[letter] = 1;

        let letterCounts = this.positionalLetterCounts.get(position);
        if (!letterCounts) {
          letterCounts = {};
          this.positionalLetterCounts.set(position, letterCounts);
        }
        if (letter in letterCounts === false) {
          letterCounts[letter] = 0;
        }
        letterCounts[letter] += 1;

        if (letterIsAlpha) {
          if (letter in this.globalAlphaLetterCounts === false) {
            this.globalAlphaLetterCounts[letter] = 0;
          }
          this.globalAlphaLetterCounts[letter] += 1;
        }
      }
    }

    //get a list of all the upper letters found in the alignment
    this.allUpperAlphaLettersInAlignmentSorted = Object.keys(
      allLettersInAlignment
    )
      .sort()
      .reduce((arr: string[], value: string) => {
        if (value.match(/[A-Z]/)) {
          // only keep letters
          arr.push(value);
        }
        return arr;
      }, []);

    //determine consensus sequence
    const consensusStats = Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        const topLetter = Object.keys(letterCounts).reduce((a, b) => {
          const aIsLetter = a.match(/[a-z]/i);
          const bIsLetter = b.match(/[a-z]/i);
          if (aIsLetter && !bIsLetter) {
            return a;
          }
          if (bIsLetter && !aIsLetter) {
            return b;
          }
          return letterCounts[a] > letterCounts[b] ? a : b;
        });
        return {
          position: position,
          letter: topLetter,
          occurrences: letterCounts[topLetter],
        };
      }
    );
    this.consensus = {
      sequence: {
        id: "consensus",
        sequence: consensusStats.map((s) => s.letter).join(""),
      },
      statistics: consensusStats,
    };

    console.log(
      "done parsing alignment. took " +
        (new Date().getTime() - start.getTime()) +
        "ms"
    );
  }

  /**
   * Is this alignment predicted to be nucleotides?
   * @returns true if it is predicted to be nucleotide sequences.
   */
  getPredictedType(): AlignmentTypes {
    return this.predictedNT
      ? AlignmentTypes.NUCLEOTIDE
      : AlignmentTypes.AMINOACID;
  }

  /**
   * Returns a default nucleotide or amino acid style for this alignment,
   * predicted from the sequence itself.
   */
  getDefaultStyle(): AminoAcidAlignmentStyle | NucleotideAlignmentStyle {
    return this.predictedNT
      ? new NucleotideAlignmentStyle()
      : new AminoAcidAlignmentStyle();
  }

  /**
   * Get a unique uuid for this alignment.
   * @returns a uuid string for this alignment
   */
  getUUID() {
    return this.uuid;
  }

  /**
   * Get the name of this alignment
   * @returns a name for the alignment - usually the filename
   */
  getName() {
    return this.name;
  }

  /**
   * Get an ace editor mode for this alignemnt. If one doesn't exist
   * then it will be defined.
   * @returns the ace editor mode for this alignment
   */
  getAceEditorMode() {
    if (!this.aceEditorMode) {
      this.aceEditorMode = "ace/mode/" + this.uuid;
      defineNewAlignmentMode(this.aceEditorMode, this);
    }
    return this.aceEditorMode;
  }

  /**
   * Get the consensus sequence
   * @returns a position ordered array of objects that contain each letter, its number of
   *          occurrences, and the actual position number
   */
  getConsensus() {
    return this.consensus;
  }

  /**
   * Get the number of gaps at each position
   * @returns an array of gap counts, ordered by position
   */
  getGapsPerColumn(): number[] {
    return Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        return !isNaN(letterCounts["-"]) ? letterCounts["-"] : 0;
      }
    );
  }

  /**
   * Get the number of gaps at a specific position
   * @returns the gap counts at the specified position
   */
  getGapCountAtColumn(position: number): number {
    const letterCounts = this.positionalLetterCounts.get(position);
    if (!letterCounts || !letterCounts["-"]) {
      return 0;
    }
    return letterCounts["-"];
  }

  /**
   * Get the total number of times each letter appears in the alignment.
   * @param normalize if true, will normalize the returned array such that
   *                  it sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   */
  getGlobalAlphaLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize) {
      return Alignment.normalizeLetterCounts(
        this.globalAlphaLetterCounts,
        validLetters
      );
    }
    return !validLetters
      ? this.globalAlphaLetterCounts
      : Object.keys(this.globalAlphaLetterCounts).reduce((acc, letter) => {
          //don't normalize, but remove invalid letters
          if (validLetters.includes(letter)) {
            acc[letter] = this.globalAlphaLetterCounts[letter];
          }
          return acc;
        }, {} as { [letter: string]: number });
  }

  /**
   * Get the length of the longest sequence
   * @returns the length of the largest sequence in this alignment
   */
  getMaxSequenceLength(): number {
    return this.getSequences().reduce((acc, seq) => {
      if (seq.sequence.length > acc) {
        acc = seq.sequence.length;
      }
      return acc;
    }, -1);
  }

  /**
   * Get the frequencies of all letters in this alignment for each position.
   * @param normalize if true, will normalize the returned letter counts for each
   *                  position such that each sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   * @returns a Map whose keys are position (ordered) and values are a dictionary
   *          with [key = letter (e.g., amino acid code)] and [value = # occurrences].
   * TODO: Test parameter use cases.
   */
  getPositionalLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize || validLetters) {
      return Array.from(this.positionalLetterCounts).reduce(
        (acc, [position, letterCounts]) => {
          acc.set(
            position,
            normalize
              ? Alignment.normalizeLetterCounts(letterCounts, validLetters) // normalize + remove invalid letters
              : Object.keys(letterCounts).reduce((acc2, letter) => {
                  //don't normalize, but remove invalid letters
                  if (validLetters!.includes(letter)) {
                    acc2[letter] = letterCounts[letter];
                  }
                  return acc2;
                }, {} as { [letter: string]: number })
          );
          return acc;
        },
        new Map<Number, { [letter: string]: number }>()
      );
    }
    return this.positionalLetterCounts;
  }

  /**
   * Get all the sequences in this alignment - ordered as they were input
   * @param sortBy specifies the order of the returned sequences. If it isn't
   *               provided, returns the sequences as they were input to create
   *               the Alignment object, e.g., usually fasta order.
   * @param forceReSort if provided, sequences are forced to be re-sorted (rather
   *                    than possibly cached)
   * @returns all sequences in this alignment
   */
  getSequences(sortBy?: SequenceSorter, forceReSort?: boolean): ISequence[] {
    sortBy = sortBy ? sortBy : SequenceSorter.INPUT;
    forceReSort = forceReSort ? forceReSort : false;
    if (
      sortBy !== SequenceSorter.INPUT &&
      (!this.sequences.has(sortBy) || forceReSort)
    ) {
      //cache not yet populated for this sort option
      this.sequences.set(
        sortBy,
        sortBy.sortFn(this.sequences.get(SequenceSorter.INPUT)!, this)
      );
    }
    return this.sequences.get(sortBy)!;
  }

  /**
   * Get a sorted list of all of the upper case letters in the alignment.
   * @returns a list of all letters in the alignment sorted alphabetically.
   */
  getAllUpperAlphaLettersInAlignmentSorted(): string[] {
    return this.allUpperAlphaLettersInAlignmentSorted;
  }

  /**
   * Get all the first sequence in the alignment, aka, target / query / first
   * @returns all sequences in this alignment
   */
  getQuerySequence(): ISequence {
    return this.querySequence;
  }

  /**
   * Set the query sequence
   * @param newQuery The new query sequence. Must already exist in the alignment.
   * @throws an error if the query sequence does not yet exist in the alignment
   *
   * IMPORTANT: Requires re-sorting
   */
  /*setQuerySequence(newQuery: ISequence) {
    const queryInAlignment = this.getSequences().find((seq) => {
      if (seq === newQuery) {
        return true;
      }
      return false;
    });
    if (queryInAlignment !== undefined) {
      this.querySequence = newQuery;

      //the resort could be really slow ... need to rethink.
      SequenceSorter.list.forEach((sortOption) => {
        //updated cached sorted sequences
        if (sortOption.isQuerySequenceDependent) {
          this.getSequences(sortOption, true);
        }
      });
      this.sequences.forEach(
        (sequences: ISequence[], sortBy: SequenceSorter) => {
          this.getSequences(sortBy, true); // force re-sorting, which updates cache
        }
      );
    } else {
      throw Error("query sequence does not exist in alignment");
    }
  }*/
}
