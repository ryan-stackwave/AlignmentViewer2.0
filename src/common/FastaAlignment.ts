import { Alignment, ISequence } from "./Alignment";

/**
 * FastaAlignment
 * This class represents a Fasta formatted multiple sequence alignment.
 */
export class FastaAlignment extends Alignment {
  /**
   * Parse a string that contains a fasta alignment.
   * @param fileName
   * @param fileContents
   */
  static fromFileContents(
    fileName: string,
    fileContents: string
  ): FastaAlignment {
    const fastaSplitCaret = fileContents.trim().split(">");
    var sequences: ISequence[] = [];
    for (var i = 0; i < fastaSplitCaret.length; i++) {
      const seqArr = fastaSplitCaret[i].split(/\r?\n/);
      if (seqArr.length > 1) {
        var seqObj = {
          id: seqArr[0],
          sequence: seqArr.slice(1).join(""),
        };
        sequences.push(seqObj);
      }
    }
    return new FastaAlignment(fileName, sequences);
  }
}
