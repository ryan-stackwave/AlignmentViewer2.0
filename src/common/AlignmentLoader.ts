/**
 * A helper class for loading alignments.
 */
import { Alignment } from "../common/Alignment";
import { FastaAlignment } from "../common/FastaAlignment";
import { StockholmAlignment } from "../common/StockholmAlignment";

export class AlignmentLoader {
  /**
   * List of all possible alignments. The order is respected when attempting
   * to parse an alignment from text / url / file.
   */
  public static AlignmentFileTypes = [StockholmAlignment, FastaAlignment];

  /**
   * Load an alignment from a URL.
   * @param url
   * @param callback will be called on successful loading of alignment
   * @param errorCallback will be called if the url fails to load or the
   *                      contents are unable to be parsed.
   * @param alignmentName explicitly set the alignment name - otherwise will
   *                      be guessed from the url.
   */
  public static async loadAlignmentFromURL(
    url: string,
    callback: (a: Alignment) => void,
    errorCallback: (e: Error[]) => void,
    alignmentName?: string
  ) {
    try {
      const f = new File(
        [await (await fetch(`${url}`)).blob()],
        alignmentName ? alignmentName : url.substring(url.lastIndexOf("/") + 1)
      );
      AlignmentLoader.loadAlignmentFromFile(f, callback, errorCallback);
    } catch (e) {
      errorCallback(e);
    }
  }

  /**
   * Load an alignment from a File.
   * @param file
   * @param callback will be called on successful loading of alignment
   * @param errorCallback will be called if the url fails to load or the
   *                      contents are unable to be parsed.
   */
  public static async loadAlignmentFromFile(
    file: File,
    callback: (a: Alignment) => void,
    errorCallback: (e: Error[]) => void
  ) {
    var reader = new FileReader();
    reader.onload = (e) => {
      try {
        callback(
          AlignmentLoader.loadAlignmentFromText(
            file.name,
            reader.result as string
          )
        );
      } catch (e) {
        errorCallback(e);
      }
    };
    reader.readAsText(file);
  }

  /**
   * Parse an alignment from text
   * @param alignmentName
   * @param text
   * @throws an excpetion if the text cannot be parsed.
   */
  public static loadAlignmentFromText(alignmentName: string, text: string) {
    const errors: Error[] = [];
    for (let i = 0; i < AlignmentLoader.AlignmentFileTypes.length; i++) {
      try {
        const toreturn = AlignmentLoader.AlignmentFileTypes[i].fromFileContents(
          alignmentName,
          text
        );
        return toreturn;
      } catch (e) {
        errors.push(e);
      }
    }
    throw errors;
  }
}
