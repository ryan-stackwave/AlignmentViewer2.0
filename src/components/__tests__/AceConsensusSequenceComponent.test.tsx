import "jest-webgl-canvas-mock";
import * as React from "react";

import { shallow, mount } from "enzyme";

import { AceConsensusSequenceComponent } from "../AceConsensusSequenceComponent";
import { Alignment } from "../../common/Alignment";
import { Ace, default as ace } from "ace-builds";
import { SequenceSorter } from "../../common/AlignmentSorter";

// Due to the runtime necessities of using styles, we need to explicitly mock out some stub data.
// https://github.com/facebook/jest/issues/3094
jest.mock("../MolecularStyles.module.scss", () => {
  return {
    aaStyBGAlpha_Default: "",
    aaStyClass_Default: "",
    aaStyColorOrder_Default: "",
    aaStyColors_Default: "",
    aaStyDesc_Default: "",
    ntStyBGAlpha_Default: "",
    ntStyClass_Default: "",
    ntStyColorOrder_Default: "",
    ntStyColors_Default: "",
    ntStyDesc_Default: "",
  };
});

describe("AceConsensusSequenceComponent", () => {
  let editor: Ace.Editor;
  let editorLoadedSpy: jest.Mock;

  beforeEach(() => {
    jest.clearAllTimers();
    jest.useFakeTimers();
    editorLoadedSpy = jest.fn((e) => {
      expect(e).not.toBeUndefined();
      expect(e).not.toBeNull();
      editor = ace.edit(e);
      expect(editor).not.toBeUndefined();
      expect(editor).not.toBeNull();
    });
  });

  it("Should render when given default props", () => {
    const wrapper = shallow(
      <AceConsensusSequenceComponent
        id="test-id"
        text={"testing"}
        fontSize={3}
        alignment={new Alignment("", [])}
        sortBy={SequenceSorter.INPUT}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Should invoke the loaded callback when the Ace editor is loaded.", () => {
    mount(
      <AceConsensusSequenceComponent
        editorLoaded={editorLoadedSpy}
        alignment={
          new Alignment("test-alignment", [
            { id: "Query", sequence: "ATGCC" },
            { id: "Psychic", sequence: "ATGCC" },
          ])
        }
        fontSize={4}
        id={"ace-msa-editor"}
        sortBy={SequenceSorter.INPUT}
      />
    );
    expect(editorLoadedSpy).toHaveBeenCalled();
    expect(editorLoadedSpy.mock.calls[0][0]).toEqual(editor);
  });
});
