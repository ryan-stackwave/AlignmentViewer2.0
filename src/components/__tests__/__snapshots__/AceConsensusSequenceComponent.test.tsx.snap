// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AceConsensusSequenceComponent Should render when given default props 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    id="test-id"
  />
</div>
`;
